packages/commonwealth/server.ts

```diff
@@ -74,6 +75,10 @@ const start = async () => {
         const databaseCleaner = new DatabaseCleaner();
         databaseCleaner.initLoop(models, config.DB.CLEAN_HOUR);
       }
+
+      setInterval(() => {
```

Comment:
I think we should check with rest of platform if this here is an ok solution(ie running a job at the server level).

Had originally thought we would be leaner on queries/RPC calls by checking individual contests on-demand in other places data was available vs a specific job at this stage

Reply:
Frankly, none of our solutions so far are okay solutions– they're hacks to get the job done faster. Platform team would disagree with the on-demand solution just as much because it performs mutations in a query.

The on-demand approach doesn't solve the UX problem in a way that Marcin/Sachin find acceptable– the rollover would be triggered in the route (background), when the user views the page. But the new contest wouldn't be projected immediately. The route would return stale data. After ~15 seconds, new contest will be persisted, now new contest will show in UI.

But this assumes that the user came back after 15 seconds and refreshed. This means that our early testers/users need to understand that behavior– or be lucky enough that a refresh happened before they landed on the page. That's non-deterministic and not a good UX.

Either we need to wait for the full rollover within the request– which is not feasible since everything is decoupled and event-driven– or we need to do it in a background loop. Or the frontend can poll the route (which is what Marcin was about to implement)– but then that's actually putting more stress on the DB because each client will poll.

The current loop solution is not too demanding because it queries contest managers, which there aren't many of yet. Also, it doesn't need to run often. Every 5 minutes is probably fine for production since the contests will last a week anyway. It doesn't actually perform RPC calls unless the contest has already ended.

Reply:
Moved the loop to the consumer rather than main app, since the consumer is guaranteed to only run a single instance and not scale horizontally. This is closer to what the proper solution will look like– decoupled from the main app.

Reply:
All makes sense to me, the consumer sounds like a good place for it for now. And given the response on the comment on the sql query requests should be limited as you said.